q2_annotate/kraken2/merge.py

```diff
@@ -0,0 +1,433 @@
+
+    # merge trees (with respect to `n_frags_assigned`, `n_frags_covered`)
+    if first_tree is None and second_tree is None:
+        merged_tree = None
+    elif first_tree is None:
+        merged_tree = second_tree
+    elif second_tree is None:
+        merged_tree = first_tree
+    else:
+        _merge_trees_recursively(from_node=first_tree, into_node=second_tree)
+        merged_tree = second_tree
+
+    unclassified_node = _merge_unclassified_nodes(
+        first_unclassified_node, second_unclassified_node
+    )
+
+    # final passover to update `perc_frags_covered`
+    if merged_tree is not None:
+        total_reads = merged_tree._kraken_data['n_frags_covered']
+        if unclassified_node is not None:
+            total_reads += unclassified_node._kraken_data['n_frags_covered']
+
+        for node in merged_tree.traverse():
+            node._kraken_data['perc_frags_covered'] = round(
+                (node._kraken_data['n_frags_covered'] / total_reads) * 100, 2
+            )
+
+    return merged_tree, unclassified_node
+
+
+def _merge_trees_recursively(
+    from_node: TreeNode, into_node: TreeNode
+) -> None:
+    '''
+    Merges the tree rooted at `from_node` into the tree rooted at `into_node`.
+    After this function completes, the `into_node` node represents the merged
+    tree.
+
+    Parameters
+    ----------
+    from_node : TreeNode
+        The root of a tree to be merged into the tree rooted by `into_node`.
+    into_node : TreeNode
+        The root of a tree into which the tree represented by `from_node`
+        will be merged.
+    '''
+    into_node._kraken_data['n_frags_covered'] += \
+        from_node._kraken_data['n_frags_covered']
+    into_node._kraken_data['n_frags_assigned'] += \
+        from_node._kraken_data['n_frags_assigned']
+
+    for child in from_node.children:
+        match = _find_node_match(child, into_node.children)
+        if match is None:
+            into_node.append(child)
+            child.parent = into_node
+        else:
+            _merge_trees_recursively(child, match)
+
+
+def _find_node_match(
+    search_node: TreeNode, children: list[TreeNode]
+) -> TreeNode | None:
+    '''
+    Searches for a match to `search_node` in `children`. The nodes' taxon ids
+    are used to define matching.
+
+    Parameters
+    ----------
+    search_node: TreeNode
+        The node for which to search in `children`.
+    children: list[TreeNode]
+        The nodes in which to search for `search_node`.
+
+    Returns
+    -------
+    TreeNode | None
+        The matching node or None if no match is found.
+
+    Raises
+    ------
+    ValueError
+        If more than one matching node is found.
+    '''
+    search_taxon_id = search_node._kraken_data['taxon_id']
+
+    matches = []
+    for child in children:
+        child_taxon_id = child._kraken_data['taxon_id']
+        if search_taxon_id == child_taxon_id:
+            matches.append(child)
```

Comment:
Someday we might change this to a dict lookup, but I don't think there's usually all that many children at any given level (especially with all of the infraclades) to make the performance win obvious.